docs/proposals/suggestion.md

```diff
@@ -0,0 +1,462 @@
+    CATEGORICAL = 4; /// Categorical type. Use "List" as string.
+}
+
+enum ObjectiveType {
+    UNKNOWN = 0; /// Undefined type and not used.
+    MINIMIZE = 1; /// Minimize
+    MAXIMIZE = 2; /// Maximize
+}
+
+message ObjectiveSpec {
+    ObjectiveType type = 1;
+    double goal = 2;
+    string objective_metric_name = 3;
+}
+
+message Trial {
+   string name = 1;
+   TrialSpec spec = 2;
+   TrialStatus status = 3;
+}
+
+message TrialSpec {
+   ParameterAssignments parameter_assignments = 2;
+   string run_spec = 3;
+}
+
+message ParameterAssignments {
+    repeated ParameterAssignment assignments = 1;
+}
+
+message ParameterAssignment {
+    string name = 1;
+    string value = 2;
+}
+
+message TrialStatus {
+   Observation observation = 4; // The best observation in logs.
+}
+
+message Observation {
+    repeated Metric metrics = 1;
+}
+
+message Metric {
+    string name = 1;
+    string value = 2;
+}
+```
+
+### Workflow
+
+![](../images/suggestion-workflow.png)
+
+When the user creates a Experiment, we will create a Suggestion for the Experiment. When the Experiment needs some suggestions, Experiment controller updates the `Suggestions`, then Suggestion controller communicates with the Suggestion to get parameter assignments and set them in Suggestion status.
+
+#### Example
+
+Now the workflow will be illustrated with an example.
+
+```yaml
+apiVersion: "kubeflow.org/v1alpha2"
+kind: Experiment
+metadata:
+  namespace: kubeflow
+  name: random-experiment
+spec:
+  parallelTrialCount: 3
+  maxTrialCount: 12
+  maxFailedTrialCount: 3
+  objective:
+    type: maximize
+    goal: 0.99
+    objectiveMetricName: Validation-accuracy
+    additionalMetricNames:
+      - accuracy
+  algorithm:
+    algorithmName: random
+  trialTemplate:
+    goTemplate:
+        rawTemplate: |-
+          apiVersion: batch/v1
+          kind: Job
+          metadata:
+            name: {{.Trial}}
+            namespace: {{.NameSpace}}
+          spec:
+            template:
+              spec:
+                containers:
+                - name: {{.Trial}}
+                  image: katib/mxnet-mnist-example
+                  command:
+                  - "python"
+                  - "/mxnet/example/image-classification/train_mnist.py"
+                  - "--batch-size=64"
+                  {{- with .HyperParameters}}
+                  {{- range .}}
+                  - "{{.Name}}={{.Value}}"
+                  {{- end}}
+                  {{- end}}
+                restartPolicy: Never
+  parameters:
+    - name: --lr
+      parameterType: double
+      feasibleSpace:
+        min: "0.01"
+        max: "0.03"
+    - name: --num-layers
+      parameterType: int
+      feasibleSpace:
+        min: "2"
+        max: "5"
+    - name: --optimizer
+      parameterType: categorical
+      feasibleSpace:
+        list:
+        - sgd
+        - adam
+        - ftrl
+```
+
+Now, we will create a Suggestion for the Experiment:
+
+```yaml
```

Comment:
Is this needed to create Suggestion with suggestions=0?

Reply:
We can make the default value to 0. Then it is not needed.

Reply:
Yes.  Since Suggestions are created during reconcile of experiments, I don't see any need in creating suggestions=0  since we know number of suggestions to be fetched.